File: ios/BleNative/ManagerWrapper.swift
import Foundation

@objc
class ManagerWrapper: NSObject {
    private let managerCache = ManagerCache()
    
    @objc
    func createCentralManager(queue: DispatchQueue, options: [String: AnyObject], callback: Callback) {
        let id = managerCache.addCentralManager(queue: queue, options: options)
        callback(createSuccessResult(data: id))
    } 
    
    @objc
    func destroyCentralManager(id: Int32) {
        managerCache.removeCentralManager(forId: id)
    }
    
    @objc
    func cancelPromise(centralManagerId: Int32, promiseId: String) {
        if let centralManager = managerCache.centralManager(forId: centralManagerId) {
            centralManager.cancelPromise(promiseId)
        }
    }
    
    // MARK: - Buffers 
    
    @objc
    public func actionOnBuffer(centralManagerId: Int32,
                               id: Int32,
                               options: [String: AnyObject], 
                               cancelOptions: [String: AnyObject], 
                               callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.actionOnBuffer(withId: id, options: options, cancelOptions: cancelOptions, callback: callback)
        }
    }
    
    @objc
    public func stopBuffer(centralManagerId: Int32, id: Int32, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.stopBuffer(withId: id, callback: callback)
        }        
    }
    
    // MARK: - State
    
    @objc
    public func getState(centralManagerId: Int32, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getState(callback: callback)
        }
    }
    
    @objc
    public func monitorState(centralManagerId: Int32, options: [String: AnyObject]?, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorState(options: options, callback: callback)
        }
    }
    
    // MARK: - State restoration
    
    @objc
    public func monitorRestoreState(centralManagerId: Int32, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorRestoreState(callback: callback)
        }
    }
    
    // MARK: - Scanning
    
    @objc
    public func scanForPeripherals(centralManagerId: Int32, filteredUUIDs: [String]?, options: [String: AnyObject]?, callback: Callback) {        
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.scanForPeripherals(filteredUUIDs, options: options, callback: callback)
        }
    }
    
    // MARK: - Name
    
    @objc
    public func getNameForPeripheral(centralManagerId: Int32, uuidString: String, callback: @escaping Callback) {        
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getNameForPeripheral(uuidString, callback: callback)
        }
    }
    
    @objc
    public func monitorPeripheralName(centralManagerId: Int32, uuidString: String, callback: @escaping Callback) {        
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorPeripheralName(uuidString, callback: callback)
        }
    }
    
    // MARK: - Read RSSI
    
    @objc
    public func readRSSIForPeripheral(centralManagerId: Int32, uuidString: String, cancelOptions: [String: AnyObject], callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.readRSSIForPeripheral(uuidString, cancelOptions: cancelOptions, callback: callback)
        }
    }
    
    // MARK: - MTU
    
    @objc
    public func requestMTUForPeripheral(centralManagerId: Int32,
                                        uuidString: String,
                                        mtu: Int,
                                        cancelOptions: [String: AnyObject],
                                        callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.requestMTUForPeripheral(uuidString, mtu: mtu, cancelOptions: cancelOptions, callback: callback)
        }
    }
    
    @objc
    public func getMTUForPeripheral(centralManagerId: Int32, uuidString: String, callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getMTUForPeripheral(uuidString, callback: callback)
        }
    }
    
    @objc
    public func monitorMTUForPeripheral(centralManagerId: Int32, uuidString: String, callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorMTUForPeripheral(uuidString, callback: callback)
        }
    }
    
    // MARK: - Device managment
    
    @objc
    public func getPeripherals(centralManagerId: Int32, deviceIdentifiers: [String], callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getPeripherals(deviceIdentifiers, callback: callback)
        }
    }
    
    @objc
    public func getConnectedPeripherals(centralManagerId: Int32, serviceUUIDs: [String], callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getConnectedPeripherals(serviceUUIDs, callback: callback)
        }
    }
    
    // MARK: - Connection managment
    
    @objc
    public func connectToPeripheral(centralManagerId: Int32, uuidString: String, options: [String: AnyObject], callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.connectToPeripheral(uuidString, options: options, callback: callback)
        }
    }
    
    @objc
    public func cancelPeripheralConnection(centralManagerId: Int32, uuidString: String, cancelOptions: [String: AnyObject], callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.cancelPeripheralConnection(uuidString, cancelOptions: cancelOptions, callback: callback)
        }
    }
    
    @objc
    public func isPeripheralConnected(centralManagerId: Int32, uuidString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.isPeripheralConnected(uuidString, callback: callback)
        }
    }
    
    @objc
    public func monitorDisconnection(centralManagerId: Int32, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorDisconnection(callback: callback)
        }
    }
    
    // MARK: - Discovery
    
    @objc
    public func discoverAllServicesAndCharacteristicsForPeripheral(centralManagerId: Int32,
                                                                   uuidString: String,
                                                                   callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.discoverAllServicesAndCharacteristicsForPeripheral(uuidString, callback: callback)
        }
    }
    
    // MARK: - Service and characteristic getters
    
    @objc
    public func getServiceForPeripheral(centralManagerId: Int32, uuidString: String, serviceUUIDString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getServiceForPeripheral(uuidString, serviceUUIDString: serviceUUIDString, callback: callback)
        }
    }
    
    @objc
    public func getServicesForPeripheral(centralManagerId: Int32, uuidString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getServicesForPeripheral(uuidString, callback: callback)
        }
    }
    
    @objc
    public func getCharacteristicForServiceByUUID(
        centralManagerId: Int32, 
        uuidString: String, 
        serviceUUIDString: String, 
        characteristicUUIDString: String, 
        callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getCharacteristicForService(uuidString, 
                                           serviceUUIDString: serviceUUIDString, 
                                           characteristicUUIDString: characteristicUUIDString,
                                           callback: callback)
        }
    }
    
    @objc
    public func getCharacteristicForService(
        centralManagerId: Int32, 
        serviceId: Int32, 
        characteristicUUIDString: String, 
        callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getCharacteristicForService(serviceId,  
                                           characteristicUUIDString: characteristicUUIDString,
                                           callback: callback)
        }
    }
    
    @objc
    public func getCharacteristicsForService(centralManagerId: Int32, serviceId: Int32, callback: Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.getCharacteristicsForService(serviceId, callback: callback)
        }
    }
    
    // MARK: - Reading
    
    @objc
    public func readBase64CharacteristicValue(centralManagerId: Int32,
                                              characteristicId: Int32,
                                              cancelOptions: [String: AnyObject],
                                              callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.readBase64CharacteristicValue(characteristicId, cancelOptions: cancelOptions, callback: callback)
        }
    }
    
    // MARK: - Writing
    
    @objc
    public func writeBase64CharacteristicValue(centralManagerId: Int32,
                                               characteristicId: Int32,
                                               valueBase64: String,
                                               response: Bool,
                                               cancelOptions: [String: AnyObject],
                                               callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.writeBase64CharacteristicValue(
                characteristicId,
                valueBase64: valueBase64,
                response: response,
                cancelOptions: cancelOptions,
                callback: callback
            )
        }
    }
    
    // MARK: - Monitoring
    
    @objc
    public func monitorBase64CharacteristicValue(centralManagerId: Int32, characteristicId: Int32, callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.monitorBase64CharacteristicValue(characteristicId, callback: callback)
        }
    }
    
    @objc
    public func isCharacteristicNotifying(centralManagerId: Int32, characteristicId: Int32, callback: @escaping Callback) {
        callOnCentralManager(centralManagerId, callback: callback) { 
            $0.isCharacteristicNotifying(characteristicId, callback: callback)
        }
    }
    
    // MARK: - Logging
    
    @objc
    public func setLogLevel(_ logLevelString: String) {
        Logger.d("ManagerWrapper setLogLevel(logLevelString: \(logLevelString))")
        Logger.changeLevel(levelString: logLevelString)
    }
    
    @objc
    public func getLogLevel(_ callback: Callback) {
        Logger.d("ManagerWrapper getLogLevel()")
        callback(createSuccessResult(data: Logger.logLevel.description))
    }
    
    
    private func callOnCentralManager(_ centralManagerId: Int32, callback: Callback, call: (CentralManager) -> ()) {
        guard let centralManager = managerCache.centralManager(forId: centralManagerId) else {
            callback(BleError.managerNotFound().asErrorResult())
            return
        }
        call(centralManager)
    }
}

File: android/src/main/java/com/polidea/blenative/ManagerWrapper.kt
package com.polidea.blenative

import android.bluetooth.BluetoothManager
import android.content.Context
import com.polidea.blenative.centralmanager.Callback
import com.polidea.blenative.centralmanager.CentralManager
import com.polidea.blenative.models.BleError
import com.polidea.blenative.utils.BleLog
import com.polidea.blenative.utils.LogLevelUtils
import com.polidea.blenative.utils.ResultUtils

class ManagerWrapper {

    private val managerCache = ManagerCache()

    fun createCentralManager(context: Context, options: Map<String, Any>, callback: Callback) {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        if (bluetoothManager == null) {
            callback(BleError.failedToCreateCentralManager().asErrorResult())
            return
        }

        val id = managerCache.addCentralManager(context, bluetoothManager)
        if (id == null) {
            callback(BleError.manyCentralManagersNotAllowed().asErrorResult())
        } else {
            callback(ResultUtils.createSuccessResult(id))
        }
    }

    fun destroyCentralManager(id: Int) {
        managerCache.removeCentralManager(id)
    }

    fun cancelPromise(centralManagerId: Int, promiseId: String) {
        managerCache.centralManager(centralManagerId)?.let {
            it.cancelPromise(promiseId)
        }
    }

    // MARK: - Buffers

    fun actionOnBuffer(centralManagerId: Int,
                       id: Int,
                       options: Map<String, Any>,
                       cancelOptions: Map<String, Any>,
                       callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.actionOnBuffer(id, options, cancelOptions, callback)
        }
    }

    fun stopBuffer(centralManagerId: Int, id: Int, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.stopBuffer(id, callback)
        }
    }

    // MARK: - State

    fun getState(centralManagerId: Int, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getState(callback)
        }
    }

    fun monitorState(centralManagerId: Int, options: Map<String, Any>?, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorState(options, callback)
        }
    }

    // MARK: - State restoration

    fun monitorRestoreState(centralManagerId: Int, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorRestoreState(callback)
        }
    }

    // MARK: - Scanning

    fun scanForPeripherals(centralManagerId: Int,
                           filteredUUIDs: Array<String>?,
                           options: Map<String, Any>,
                           callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.scanForPeripherals(filteredUUIDs, options, callback)
        }
    }

    // MARK:- Name

    fun getNameForPeripheral(centralManagerId: Int, uuidString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getNameForPeripheral(uuidString, callback)
        }
    }

    fun monitorPeripheralName(centralManagerId: Int, uuidString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorPeripheralName(uuidString, callback)
        }
    }

    // MARK: - Read RSSI

    fun readRSSIForPeripheral(centralManagerId: Int,
                              uuidString: String,
                              cancelOptions: Map<String, Any>,
                              callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.readRSSIForPeripheral(uuidString, cancelOptions, callback)
        }
    }

    // MARK: - MTU

    fun requestMTUForPeripheral(centralManagerId: Int,
                                uuidString: String,
                                mtu: Int,
                                cancelOptions: Map<String, Any>,
                                callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.requestMTUForPeripheral(uuidString, mtu, cancelOptions, callback)
        }
    }

    fun getMTUForPeripheral(centralManagerId: Int,
                            uuidString: String,
                            callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getMTUForPeripheral(uuidString, callback)
        }
    }

    fun monitorMTUForPeripheral(centralManagerId: Int,
                                uuidString: String,
                                callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorMTUForPeripheral(uuidString, callback)
        }
    }

    // MARK: - Device managment

    fun getPeripherals(centralManagerId: Int,
                       deviceIdentifiers: Array<String>,
                       callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getPeripherals(deviceIdentifiers, callback)
        }
    }

    fun getConnectedPeripherals(centralManagerId: Int,
                                serviceUUIDs: Array<String>,
                                callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getConnectedPeripherals(serviceUUIDs, callback)
        }
    }

    // MARK: - Connection managment

    fun connectToPeripheral(centralManagerId: Int,
                            uuidString: String,
                            options: Map<String, Any>,
                            callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.connectToPeripheral(uuidString, options, callback)
        }
    }

    fun cancelPeripheralConnection(centralManagerId: Int,
                                   uuidString: String,
                                   cancelOptions: Map<String, Any>,
                                   callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.cancelPeripheralConnection(uuidString, cancelOptions, callback)
        }
    }

    fun isPeripheralConnected(centralManagerId: Int, uuidString: String, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.isPeripheralConnected(uuidString, callback)
        }
    }

    fun monitorDisconnection(centralManagerId: Int, callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorDisconnection(callback)
        }
    }

    // MARK: - Discovery

    fun discoverAllServicesAndCharacteristicsForPeripheral(centralManagerId: Int,
                                                           uuidString: String,
                                                           callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.discoverAllServicesAndCharacteristicsForPeripheral(uuidString, callback)
        }
    }

    // MARK: - Service and characteristic getters

    fun getServiceForPeripheral(centralManagerId: Int,
                                uuidString: String,
                                serviceUUIDString: String,
                                callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getServiceForPeripheral(uuidString, serviceUUIDString, callback)
        }
    }

    fun getServicesForPeripheral(centralManagerId: Int,
                                 uuidString: String,
                                 callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getServicesForPeripheral(uuidString, callback)
        }
    }

    fun getCharacteristicForServiceByUUID(centralManagerId: Int,
                                          uuidString: String,
                                          serviceUUIDString: String,
                                          characteristicUUIDString: String,
                                          callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getCharacteristicForService(uuidString,
                    serviceUUIDString,
                    characteristicUUIDString,
                    callback)
        }
    }

    fun getCharacteristicForService(centralManagerId: Int,
                                    serviceId: Int,
                                    characteristicUUIDString: String,
                                    callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getCharacteristicForService(serviceId,
                    characteristicUUIDString,
                    callback)
        }
    }

    fun getCharacteristicsForService(centralManagerId: Int,
                                     serviceId: Int,
                                     callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.getCharacteristicsForService(serviceId, callback)
        }
    }

    // MARK: - Reading

    fun readBase64CharacteristicValue(centralManagerId: Int,
                                      characteristicId: Int,
                                      cancelOptions: Map<String, Any>,
                                      callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.readBase64CharacteristicValue(characteristicId, cancelOptions, callback)
        }
    }

    // MARK: - Writing

    fun writeBase64CharacteristicValue(centralManagerId: Int,
                                       characteristicId: Int,
                                       valueBase64: String,
                                       response: Boolean,
                                       cancelOptions: Map<String, Any>,
                                       callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.writeBase64CharacteristicValue(
                    characteristicId,
                    valueBase64,
                    response,
                    cancelOptions,
                    callback
            )
        }
    }

    // MARK: - Monitoring

    fun monitorBase64CharacteristicValue(centralManagerId: Int,
                                         characteristicId: Int,
                                         callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.monitorBase64CharacteristicValue(characteristicId, callback)
        }
    }

    fun isCharacteristicNotifying(centralManagerId: Int,
                                  characteristicId: Int,
                                  callback: Callback) {
        callOnCentralManager(centralManagerId, callback) {
            it.isCharacteristicNotifying(characteristicId, callback)
        }
    }

    // MARK: - Logging

    fun setLogLevel(logLevelString: String) {
        BleLog.d("ManagerWrapper setLogLevel($logLevelString)")
        val logLevel = LogLevelUtils.fromLevelString(logLevelString)
        BleLog.setLogLevel(logLevel)
    }

    fun getLogLevel(callback: Callback) {
        BleLog.d("ManagerWrapper getLogLevel()")
        val levelDescription = LogLevelUtils.levelDescription(BleLog.getLogLevel())
        callback(ResultUtils.createSuccessResult(levelDescription))
    }

    private fun callOnCentralManager(centralManagerId: Int, callback: Callback, call: (CentralManager) -> Unit) {
        val centralManager = managerCache.centralManager(centralManagerId)
        if (centralManager == null) {
            callback(BleError.managerNotFound().asErrorResult())
            return
        }
        call(centralManager)
    }
}